LCOE calculations with parasitic load costs in numerator, Energy sales in denominator

mjprilliman opened the pull request:
-Updated LCOE calculation in all financial models
-Parasitic load operating cost based on PPA price or retail rate for FOM
-Parasitic load operating cost based on hourly prices for BTM
-Variable generation O&M calculated based on energy of value (FOM-sales, BTM-to grid or to load) not net energy
-LCOS updated with same LCOE methods for calculation of cost of grid charging
-Closes #546 outside of larger powerflow changes

cpaulgilman commented:
@mjprilliman What branch of SAM is best to test this with?

mjprilliman commented:
> @mjprilliman What branch of SAM is best to test this with?

Identical branch name in SAM. https://github.com/NREL/SAM/pull/749

cpaulgilman reviewed (commented):
I'm not sure how to evaluate these changes to the LCOE, but it seems to be working as intended.

I added electricity purchases to cash flow for Detailed PV / PPA configs. Not sure if it should also be added for other technologies.

SAM crashes for CSP systems: 

1. Create CSP / Single Owner case.
2. Go to Electricity Purchases page
3. Choose **Retail electricity rate(s) option**

mjprilliman commented:
> I'm not sure how to evaluate these changes to the LCOE, but it seems to be working as intended.
> 
> I added electricity purchases to cash flow for Detailed PV / PPA configs. Not sure if it should also be added for other technologies.
> 
> SAM crashes for CSP systems:
> 
> 1. Create CSP / Single Owner case.
> 2. Go to Electricity Purchases page
> 3. Choose **Retail electricity rate(s) option**

Either through code merges or a mistake on my part en_electricity_rates was calculated in the equations and updated in the callbacks, which was causing intermittent crashing for every configuration. Hoping the most recent push will fix the crashing, running the test script now.

cpaulgilman commented:
If we are making the electricity purchases option (use PPA or market prices / use retail electricity rates) available for all FOM configurations in the UI, then the FOM financial models need to be updated to calculate cf_energy_purchases_value and cf_utility_bill and include it in the revenue and operating expenses calculation as appropriate. 

As of now, this is handled for single owner, but not all equity partnership flip, leveraged partnership flip, sale leaseback, or merchant plant.

I have a modification to cashflow.lk that adds cf_energy_purchases_value and cf_utility_bill to the cash flow in the UI that I can commit to this branch if we are taking that path. This change would also require updating all of the send-to-Excel with equations workbooks.

brtietz commented on ssc/cmod_equpartflip.cpp:
```diff
@@ -915,6 +919,9 @@ class cm_equpartflip : public compute_module
 		int i = 0;
 
+        if (is_assigned("en_electricity_rates") && as_number("en_electricity_rates") == 0 && as_number("ppa_soln_mode") == 0)
+            throw exec_error("singleowner", "PPA price from which to calculate parasitic load costs is not specified. Check inputs for Revenue and Electricity Purchases.");
```
Please change the cmod name in the error message

brtietz commented on ssc/cmod_cashloan.cpp:
```diff
@@ -990,7 +1104,7 @@ class cm_cashloan : public compute_module
 		}
 
-		double npv_energy_nom = npv( CF_energy_net, nyears, nom_discount_rate );
+		double npv_energy_nom = npv( CF_energy_sales, nyears, nom_discount_rate ) + npv(CF_parasitic_cost, nyears, nom_discount_rate);
```
I thought CF_parasitic_cost went in the numerator. This is the energy term in the denominator of the nominal LCOE.

brtietz commented on ssc/cmod_host_developer.cpp:
```diff
@@ -2657,16 +2707,55 @@ class cm_host_developer : public compute_module
 
-	double host_npv_energy_nom = npv(CF_energy_net, nyears, host_nom_discount_rate);
+	double host_npv_energy_nom = npv(CF_energy_sales, nyears, host_nom_discount_rate);
 	double host_lcoe_nom = -(cf.at(CF_after_tax_net_equity_cost_flow, 0) + npv(CF_after_tax_net_equity_cost_flow, nyears, host_nom_discount_rate)) * 100;
```
I see where the parasitics are included in the real LCOE, but what about nominal?

brtietz reviewed (changes requested):
My code review found a few typos worth fixing.

It also looks like lnte in Third Party uses CF_energy_net in the denominator - should this be updated as well?

mjprilliman commented:
> If we are making the electricity purchases option (use PPA or market prices / use retail electricity rates) available for all FOM configurations in the UI, then the FOM financial models need to be updated to calculate cf_energy_purchases_value and cf_utility_bill and include it in the revenue and operating expenses calculation as appropriate.
> 
> As of now, this is handled for single owner, but not all equity partnership flip, leveraged partnership flip, sale leaseback, or merchant plant.
> 
> I have a modification to cashflow.lk that adds cf_energy_purchases_value and cf_utility_bill to the cash flow in the UI that I can commit to this branch if we are taking that path. This change would also require updating all of the send-to-Excel with equations workbooks.

I'm cleaning up the cash flow outputs in the other FOM financial models (as well as fixing the issues Brian found). So it would now be possible to have the energy purchases and utility bill in the cash flow and Send to Excel for each. I'm not opposed to eliminating the option for some FOM configurations, but how do we decide which configs have that option and which do not?

cpaulgilman commented:
> > If we are making the electricity purchases option (use PPA or market prices / use retail electricity rates) available for all FOM configurations in the UI, then the FOM financial models need to be updated to calculate cf_energy_purchases_value and cf_utility_bill and include it in the revenue and operating expenses calculation as appropriate.
> > As of now, this is handled for single owner, but not all equity partnership flip, leveraged partnership flip, sale leaseback, or merchant plant.
> > I have a modification to cashflow.lk that adds cf_energy_purchases_value and cf_utility_bill to the cash flow in the UI that I can commit to this branch if we are taking that path. This change would also require updating all of the send-to-Excel with equations workbooks.
> 
> I'm cleaning up the cash flow outputs in the other FOM financial models (as well as fixing the issues Brian found). So it would now be possible to have the energy purchases and utility bill in the cash flow and Send to Excel for each. I'm not opposed to eliminating the option for some FOM configurations, but how do we decide which configs have that option and which do not?

I think we should make the option available for all FOM configurations. I'll commit my cashflow.lk that includes those outputs in the cash flow for all FOM configs so you can test with your changes to the financial models.

brtietz commented on ssc/cmod_cashloan.cpp:
```diff
@@ -976,11 +1022,79 @@ class cm_cashloan : public compute_module
+                                monthly_e_fromgrid[m-1] += year1_hourly_e_from_grid[util::hour_of_year(m, d, h) * n_steps_per_hour + n];
+                                monthly_gen_purchases[(a - 1) * 12 + m-1] += -gen_purchases[(size_t(a) - 1) * 8760 * n_steps_per_hour + n_steps_per_hour * util::hour_of_year(m, d, h) + n];
+                                if (year1_hourly_e_from_grid[h] != 0.0) {
+                                    //cf.at(CF_charging_cost_grid_month, a) += monthly_grid_to_batt[m] / (monthly_grid_to_batt[m] + monthly_grid_to_load[m]) * monthly_energy_charge[m] * charged_grid[a] / charged_grid[1] * cf.at(CF_util_escal_rate, a);
```
Do you mind removing the commented out lines prior to the merge?

brtietz commented on ssc/cmod_equpartflip.cpp:
```diff
@@ -1149,6 +1187,68 @@ class cm_equpartflip : public compute_module
+                if (!ppa_purchases) {
+                    cf.at(CF_energy_purchases_value, i) = 0.0;
+                    // Recompute this variable because the ppa_gen values (hourly_net) were all positve until now 
+                    //cf.at(CF_energy_net, i) = cf.at(CF_energy_sales, i) + cf.at(CF_energy_purchases, i); // Adding a positive and negative number
```
More commented out code - looks like the comment could also be removed?

brtietz commented on ssc/cmod_equpartflip.cpp:
```diff
@@ -1059,62 +1067,92 @@ class cm_equpartflip : public compute_module
+            ssc_number_t* ub_arr;
+            ub_arr = as_array("utility_bill_w_sys", &ub_count);
+            if (ub_count != (size_t)(nyears + 1))
+                throw exec_error("singleowner", util::format("utility bill years (%d) not equal to analysis period years (%d).", (int)ub_count, nyears));
```
Issue with cmod name in exception

brtietz commented on ssc/cmod_singleowner.cpp:
```diff
@@ -2979,27 +3008,28 @@ class cm_singleowner : public compute_module
+            + cf.at(CF_reserve_interest, i)
+            - cf.at(CF_disbursement_debtservice, i) // note sign is negative for positive disbursement
+            - cf.at(CF_disbursement_om, i) // note sign is negative for positive disbursement
+            //+ cf.at(CF_energy_purchases_value, i);
```
Why is this commented out for singleowner but not the other compute modules?

brtietz reviewed (commented):
This is close. I think taking two steps for cleanliness would help a lot:

(1) Do a search for <//cf> and see if the line is supposed to be commented out. If yes, delete it and associated comments.

(2) Do a search for <throw exec_error("singleowner"> in compute modules that are not singleowner and update the cmod name appropriately.

The brackets in the above are just to indicate the search term without quotes (since the singleowner search ends with a quote)

brtietz reviewed (approved):
Great work! Thanks for doing the comment and exception cleanup. I think this is ready to go once the SAM side of things is ready to go.
